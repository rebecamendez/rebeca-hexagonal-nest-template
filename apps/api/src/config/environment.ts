import { Expose, Transform, plainToClass } from 'class-transformer';
import { IsBoolean, IsIn, IsNumber, IsOptional, IsString, ValidationError, validateSync } from 'class-validator';

import { Logger } from '@nestjs/common';

export class Environment {
  @Expose()
  @IsOptional()
  @IsNumber()
  @Transform(({ value }) => Number(value))
  public readonly PORT: number = 3000;

  @Expose()
  @IsOptional()
  @IsString()
  public readonly DB_HOST: string = 'localhost';

  @Expose()
  @IsOptional()
  @IsNumber()
  @Transform(({ value }) => Number(value))
  public readonly DB_PORT: number = 5432;

  @Expose()
  @IsOptional()
  @IsString()
  public readonly DB_USER: string = 'rebeca-hexagonal-nest-template';

  @Expose()
  @IsOptional()
  @IsString()
  public readonly DB_PASSWORD: string = 'rebeca-hexagonal-nest-template';

  @Expose()
  @IsOptional()
  @IsString()
  public readonly DB_DATABASE: string = 'rebeca-hexagonal-nest-template';

  @Expose()
  @IsOptional()
  @IsBoolean()
  @Transform(({ value }) => value === 'true')
  public readonly DB_ENABLE_LOG: boolean = false;

  @Expose()
  @IsOptional()
  @IsBoolean()
  @Transform(({ value }) => value === 'true')
  public readonly ENABLE_OPENAPI: boolean = false;

  @IsIn(['LOCAL', 'PROD'])
  public readonly ENVIRONMENT: 'LOCAL' | 'PROD' = 'LOCAL';

  // -- Static Methods --
  public static validate(config: Record<string, unknown>): Environment {
    const validatedConfig = plainToClass(Environment, config, {
      enableImplicitConversion: false,
      excludeExtraneousValues: true,
      exposeDefaultValues: true,
      exposeUnsetFields: false
    });

    const errors = validateSync(validatedConfig, { skipMissingProperties: false, whitelist: true, forbidNonWhitelisted: true });

    if (errors.length > 0) {
      this.formatValidationErrors(errors);
      throw new Error('Environment validation failed');
    }

    Logger.log('[environment] ✅ Variables validated successfully');
    return validatedConfig;
  }

  // -- Private Methods --
  private static formatValidationErrors(errors: ValidationError[]): void {
    Logger.error('[environment] ❌ Error validating environment variables:');
    errors.forEach((error) => {
      Logger.error(`   - ${error.property}: ${Object.values(error.constraints || {}).join(', ')}`);
    });
  }
}
