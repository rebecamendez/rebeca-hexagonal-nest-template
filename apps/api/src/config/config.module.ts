import { Global, Module } from '@nestjs/common';
import { ConfigModule as NestConfigModule } from '@nestjs/config';

import { Environment } from './environment';

@Global()
@Module({
  imports: [
    NestConfigModule.forRoot({
      validate: (config: Record<string, unknown>) => Environment.validate(config),
      isGlobal: true,
      cache: true,
      envFilePath: '.env'
    })
  ]
})
export class ConfigModule {}
