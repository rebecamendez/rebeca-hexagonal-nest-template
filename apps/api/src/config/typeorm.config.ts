import { DataSource } from 'typeorm';

import { Environment } from './environment';

const env = new Environment();

export default new DataSource({
  type: 'postgres',
  host: env.DB_HOST,
  port: env.DB_PORT,
  username: env.DB_USER,
  password: env.DB_PASSWORD,
  database: env.DB_DATABASE,
  entities: ['src/**/*.entity.ts'],
  migrations: ['src/migrations/*.ts'],
  synchronize: false
});
