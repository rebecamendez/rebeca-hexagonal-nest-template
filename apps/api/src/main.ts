import { WinstonModule } from 'nest-winston';
import { winstonLoggerOptions } from 'utils/logger';

import { Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config/dist/config.service';
import { NestFactory } from '@nestjs/core';

import { AppModule } from './app.module';
import { setupNestApp } from './main.setup';

async function bootstrap(): Promise<void> {
  const app = await NestFactory.create(AppModule, {
    cors: true,
    logger: WinstonModule.createLogger(winstonLoggerOptions),
    rawBody: true
  });

  const configService = app.get(ConfigService);
  const port = configService.getOrThrow<number>('PORT');
  const enableOpenapi = configService.getOrThrow<boolean>('ENABLE_OPENAPI');

  setupNestApp(app, port, enableOpenapi);

  await app.listen(port);

  Logger.log(`[main] ðŸš€ Application is running on: http://localhost:${port}`);
}

void bootstrap();
