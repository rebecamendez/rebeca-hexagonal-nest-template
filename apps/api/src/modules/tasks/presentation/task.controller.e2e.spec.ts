/* eslint-disable no-process-env */
import { INestApplication } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { setupNestApp } from 'main.setup';
import { TaskEntityBuilder } from 'modules/shared/database/tests/task.entity.builder';
import request from 'supertest';
import { StartedTestContainer } from 'testcontainers';
import { containerSetup } from 'tests/test-containers.setup';
import { DataSource } from 'typeorm';
import { AppModule } from '../../../app.module';

describe('Task Endpoints (e2e)', () => {
  let app: INestApplication;
  let container: StartedTestContainer;
  let dataSource: DataSource;

  const prepareScenario = async (dataSource: DataSource): Promise<void> => {
     await new TaskEntityBuilder().mock().save(dataSource);
  };

  beforeAll(async () => {

    const databaseName = 'task-endpoints-e2e-test';
    ({ container, dataSource } = await containerSetup(databaseName));

    await prepareScenario(dataSource);

    const mockedConfig = jest.fn((key: string) => {
      const map: Record<string, string | undefined> = process.env;
      switch (key) {
        case 'DB_HOST':
          return container.getHost();
        case 'DB_PORT':
          return container.getMappedPort(5432);
        case 'DB_USER':
          return 'the-user';
        case 'DB_PASSWORD':
          return 'the-password';
        case 'DB_DATABASE':
          return databaseName;
      }
      return map[key];
    });

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule]
    })
      .overrideProvider(ConfigService)
      .useValue({ get: mockedConfig, getOrThrow: mockedConfig })
      .compile();

    app = moduleFixture.createNestApplication();
    setupNestApp(app, 3000, false); // https://stackoverflow.com/questions/59355841/how-to-apply-global-pipes-during-e2e-tests
    await app.init();
  });

  afterAll(async () => {
    if (app) await app.close();
    if (container) await container.stop();
  });

  describe('when retrieving tasks', () => {
    it('should return a list of tasks', async () => {
      const response = await request(app.getHttpServer())
        .get('/tasks')
        .send();

      expect(response.status).toBe(200);
      expect(response.body).toMatchSnapshot();
    });
  });
});
