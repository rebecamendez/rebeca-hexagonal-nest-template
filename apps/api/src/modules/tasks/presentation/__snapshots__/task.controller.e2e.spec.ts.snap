// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Endpoints (e2e) when retrieving tasks should return a list of tasks 1`] = `
[
  {
    "description": "description",
    "id": 1,
    "title": "title",
  },
]
`;
